### PR TITLE
CRM-1578-[DMS] - Duplicates of CH Funds Showing Under Administer > Funds

### PR DESCRIPTION
**Pull request by CH-Rajal**
The following code will handle duplicate CH fund values and eliminate the possibility to create a new duplicate CH fund.

Example:

If in the existing system there are multiple CH funds under the same label but with different values and associated with different contributions

Data table : civicrm_option_value
+------+-----------------+-------------+------------+----------+--------+------------+--------+-------------+-------------+-----------+--------------+-----------+---------------+------+-------+-------------+-------------------+-------------+-------------------+-------------+-------------------+
| id   | option_group_id | value       | name       | grouping | filter | is_default | weight | is_optgroup | is_reserved | is_active | component_id | domain_id | visibility_id | icon | color | label_en_US | description_en_US | label_fr_CA | description_fr_CA | label_en_CA | description_en_CA |
+------+-----------------+-------------+------------+----------+--------+------------+--------+-------------+-------------+-----------+--------------+-----------+---------------+------+-------+-------------+-------------------+-------------+-------------------+-------------+-------------------+
| 1356 |             116 | CH+00111111 | Rajal_Test | NULL     |      0 |          0 |    115 |           0 |           0 |         1 |         NULL |      NULL |          NULL | NULL | NULL  | Rajal Test  | NULL              | Rajal Test  | NULL              | Rajal Test  | NULL              |
| 1359 |             116 | CH+00222222 | Rajal_Test | NULL     |      0 |          0 |    116 |           0 |           0 |         1 |         NULL |      NULL |          NULL | NULL | NULL  | Rajal Test  | NULL              | Rajal Test  | NULL              | Rajal Test  | NULL              |
+------+-----------------+-------------+------------+----------+--------+------------+--------+-------------+-------------+-----------+--------------+-----------+---------------+------+-------+-------------+-------------------+-------------+-------------------+-------------+-------------------+
Data table: civicrm_option_value_ch
+-----+-----------------+---------------+-------------------+------------------+-----------+
| id  | option_group_id | value         | financial_type_id | is_enabled_in_ch | parent_id |
+-----+-----------------+---------------+-------------------+------------------+-----------+
| 231 |             116 | CH+00111111   |               1   |                0 |       NULL|
| 232 |             116 | CH+00222222   |               1   |                0 |       NULL|
+-----+-----------------+---------------+-------------------+------------------+-----------+

To prevent creating a new CH Fund value under the name 'Rajal_Test' and a different value CH+00333333, add this data under civicrm_option_value_ch, table and for that entry parent_id should be 231. That way we can track associated CH fund with the same name under different values.

To manage historical data, will remove all duplicate ch funds which are not associated with the contribution and later when sync will happen it will sync all CH funds values again and this time it will follow the parent_id association logic.

(1)CRM/Chfunds/DAO/OptionValueCH.php

Define field value for 'parent_id' column .Inserted new column 'parent_id' for data table civicrm_option_value_ch

(2)CRM/Chfunds/Form/CHFunds.php

From DMS system when charity admin reassigns CH fund(update value of financial_type_id), code will ensure that along with that CH fund(parent fund), related child ch funds also gets associate with correct financial_type_id.

(3)CRM/Chfunds/Upgrader.php

Written two upgrader functions for the following reasons.
(1) alter civicrm_option_value_ch table to insert'parent_id' column.
(2) code to manage historical duplicate CH funds and associated contributions.

(4)CRM/Chfunds/Utils.php

While syncing contributions from CH to DMS, retrieve the proper value of option value which can associate with contributions.

(5)CRM/OptionValue/CreateAPIWrapper.php

While creating a new CH fund option value, the code will check if the existing DB option value table contains the same label value or not, if yes then will simply return the existing option value data and create a new value of option_value_ch table. If not then create new values for option_values and option_values_ch tabels.

(6)CRM/OptionValue/DeleteAPIWrapper.php 

When ch fund value gets removed from the option_value table and option_value_ch table, the system will make sure the associated child option_value_ch values also get removed.

(7)api/v3/CHContribution/Create.php

Added condition while creating new CH contribution.

(8)chfunds.php

When super Admin, updates option value for CH funds from settings, when we re-assign one CH fund to any other fund, identify  OptionValueCH which has the parent ID of CH fund which is going to be reassigned

### CODE REVIEW

File: CRM/Chfunds/Upgrader.php [hunk 1]
```diff
@@ -96,6 +98,132 @@ public function upgrade_1700() {
...
+    //Adding a new column 'parent_id' for dataTable `civicrm_option_value_ch`
+    $sql = "ALTER TABLE `civicrm_option_value_ch` ADD COLUMN `parent_id` INT NULL DEFAULT NULL COMMENT 'parent id of CH option value' AFTER is_enabled_in_ch";
+    CRM_Core_DAO::executeQuery($sql);
+    return TRUE;
+
+  }
+
+  public function upgrade_1810() {
+    $this->ctx->log->info('CRM-1578-[DMS] -Manage Historical CH Funds data');
+    //Get option Group ID value for 'ch_fund'
```
Comment (celinenicolas) [inline CRM/Chfunds/Upgrader.php]: this code can be simplified with a single API chained call

 ```suggestion
  // Get funds with duplicate names
  $optionValues = civicrm_api4('OptionGroup', 'get', [
    'select' => [
      'id',
    ],
    'where' => [
      ['name', '=', 'activity_type'],
    ],
    'limit' => 1,
    'chain' => [
      'funds' => [
        'OptionValue', 'get', [
          'select' => [
            'name', 
            'COUNT(name) AS count'
          ], 
          'groupBy' => [
            'name'
          ], 
          'having' => [
            ['count', '>', 1]
          ], 
          'where' => [
            ['option_group_id', '=', '$id']
          ]
        ]
      ],
    ],
  ]);

  if ( count($optionValues) > 0 && count($optionValues[0]['funds']) > 0 {
    // loop through $optionValues[0]['funds']
  }
```

Comment (celinenicolas) [reply CRM/Chfunds/Upgrader.php]: api test url: https://test2-dms.canadahelps.org/dms/api4#/explorer/OptionGroup/get?where=%5B%5B%22name%22,%22%3D%22,%22ch_fund%22%5D%5D&chain=%5B%5B%22name_me_0%22,%5B%22OptionValue%22,%22get%22,%22%7Bselect:%20%5Bname,%20COUNT(name)%20AS%20count%5D,%20groupBy:%20%5Bname%5D,%20having:%20%5B%5B'count',%20'%3E',%201%5D%5D,%20where:%20%5B%5Boption_group_id,%20%3D,%20$id%5D%5D%7D%22,%22%22%5D%5D%5D&limit=1&select=%5B%22id%22%5D

File: CRM/Chfunds/Utils.php [hunk 2]
```diff
@@ -11,6 +11,32 @@ public static function getFinancialTypeByCHFund($chFundID) {
 
     return CRM_Utils_Array::value('financial_type_id', $values);
   }
+  //CRM-1578- create function to get orginal option value for child CH fund which can associate with contribution
```
Comment (celinenicolas) [inline CRM/Chfunds/Utils.php:14]: same issue here. 
those 3 api calls could be chained into a single call

File: CRM/OptionValue/CreateAPIWrapper.php [hunk 3]
```diff
@@ -31,10 +31,57 @@ public function fromApiInput($apiRequest) {
           $params['id'] = $OVCHid;
         }
         else {
-          // always enusre that whenever a new ch_fund optionValue is created its always reserved so that it cant be deleted from UI
-          $apiRequest['params']['is_reserved'] = 1;
-          // Allocate newly created CH Fund option to 'General Fund'
-          $params['financial_type_id'] = civicrm_api3('FinancialType', 'getvalue', ['name' => 'General Fund', 'return' => 'id']);
+          //CRM-1578 identify duplicate option value for CH fund based on name
```
Comment (celinenicolas) [inline CRM/OptionValue/CreateAPIWrapper.php:34]: same issue here.
those api calls could be chained into a single call

**Review by celinenicolas — CHANGES_REQUESTED** [review]

**Review by celinenicolas — CHANGES_REQUESTED** [review]
We really have to reduce the number of API calls.
Considering this is an upgrader function, and we're already struggling with simple ones, I'm not at all confident we could run this upgrader function.

At the very minimum, which is if we only have 1 duplicate, we're executing no less than 5 api calls. Now, let's imagine we have 5 duplicates, that's like 30 api calls. We're gonna overload the servers.

Using api4 (when we can) + some code optimization, we should be able to reduce those calls to 3+.

We can discuss this on Monday.

**Comment by CH-Rajal** [comment]
For the Latest Commit following are Major logic changes we have considered.

(1) Earlier from CH to the DMS end they were not sending any flags to display whether that specific fund is activated or deleted on CH end. With my discussion with Kenny, I learned they are sending active flag params to indicate whether that specific fund with a duplicate name is active or not.
Based on that while syncing CH funds from CH to DMS we will get the idea out of (2 or 3 or any number of duplicate funds with the same name which fund is active and which fund is not active) and based on that we can take reference of the active fund and make it as parent id reference for all other duplicate funds. re-assign all contributions from duplicate funds to those, and delete the duplicate entries.

In the Upgrader function, we have replaced API calls with SQL queries.If both the fund name and GL Account name DO match, than use the latest Fund and GL Account as the "parent", re-assign all contributions from duplicate funds to those, and delete the duplicate entries. For CH Funds, we will keep some record of the duplicate for syncing, and include the notion of a "parent" fund.

**Review by celinenicolas — APPROVED** [review]

**Pull request merged**